src/main/java/org/tikv/common/log/SlowLogImpl.java

```diff
@@ -0,0 +1,97 @@
+/*
+ *
+ * Copyright 2021 PingCAP, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.tikv.common.log;
+
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SlowLogImpl implements SlowLog {
+  private static final Logger logger = LoggerFactory.getLogger(SlowLogImpl.class);
+
+  public static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("HH:mm:ss.SSS");
+
+  private final List<SlowLogTask> slowLogTasks = new ArrayList<>();
+
+  private final long startMS;
+  private final Map<String, String> properties;
+
+  public SlowLogImpl(Map<String, String> properties) {
+    this.startMS = System.currentTimeMillis();
+    this.properties = properties;
+  }
+
+  @Override
+  public synchronized SlowLogTask start(String taskName) {
+    SlowLogTask slowLogTask = new SlowLogTaskImpl(taskName);
+    slowLogTasks.add(slowLogTask);
+    slowLogTask.start();
+    return slowLogTask;
+  }
+
+  @Override
+  public void logSlowLog(boolean pretty) {
+    logger.warn(getSlowLogString(pretty));
+  }
+
+  private String getSlowLogString(boolean pretty) {
```

Comment:
do we have other methods to build the JSON without hand-coding the JSON string?

Reply:
like this: https://docs.oracle.com/javaee/7/api/javax/json/JsonObjectBuilder.html